### ISSUE TITLE
deepxde/gradients.py and unused gradients

### ISSUE THREAD

**Issue by smao-astro** [lululxvi/deepxde#163]
Hi Lu,

I have read the code in `deepxde/gradients.py`, which is clear and well written, but I have some questions here:

For example my PDE have 2 input (x1, x2) and 1 output y:

1. If dy/dx1 is needed while dy/dx2 is not (explicitly and implicitly), do you think `dde.grad.jacobian(y, x, i=0, j=0)` evaluates dy/dx2 as well?
2. If d^2y/dx1^2 is needed while d^2y/dx1dx2 is not (explicitly and implicitly), do you think `dde.grad.hessian(y, x, i=0, j=0)` evaluates d^2y/dx1dx2 as well?

**Comment by lululxvi** [comment]
Good questions. 

1. Yes. `dde.grad.jacobian(y, x, i=0, j=0)` also evaluates dy/dx2, because I use `tf.gradients(y, x)`, which computes both dy/dx1 and dy/dx2. If we only need dy/dx1, maybe we can use the code:
```python
x1 = X[:, 0:1]
dy_dx1 = tf.gradients(y, x1)
```
But I didn't test it and am not sure if this works. Also, because we use back-propagation (see Section 2.2 in [DeepXDE paper](https://arxiv.org/pdf/1907.04502.pdf)), actually the cost of `tf.gradients(y, x)` and `tf.gradient(y, x1)` is almost the same. There is no extra cost. In summary, if we need both dy/dx1 and dy/dx2, then `tf.gradients(y, x)` is the best solution, which computes these two gradients together in one back-propagation; even if we only need one of the gradients, the cost is the same as we compute all the gradients in one back-propagation. Or we could say dy/dx2 is obtained for free.

2. To compute d^2y/dx1^2, we first compute dy/dx1 and then compute d(dy/dx1)/dx1. When we compute d(dy/dx1)/dx1, we also compute d(dy/dx1)/dx2. But the same as Part 1, d(dy/dx1)/dx2 is obtained for free when we compute d(dy/dx1)/dx1.

So it is not a problem. There is no extra computation cost.

**Comment by smao-astro** [comment]
I see, thank you very much for your prompt reply!

**Comment by smao-astro** [comment]
Hi Lu,

I have another might related question:

Let's say now I have one input x and one NN output y, and I want to compute dy/dx and d(sin(x) * y)/dx, I know that in such case  the second one I can not **directly** fetch the value stored in dde.gradients.jacobian(y, x, i=0, j=0) that precomputed by dy/dx. But the question is that, can dde.gradients.jacobian(sin(x) * y, x, i=0, j=0) have a way to reuse the intermediate partial derivative value of the chain rule (maybe they are stored in the nodes of the graph? sorry I do not know how TF managed that and did not find material online that give the answer)?

For example, I am curious whether `tf.gradients(tf.sin(x[:, 0]) * y, x)[0][:, 0]` and `tf.gradients(tf.sin(x[:, 0]), x)[0][:, 0] * y + tf.sin(x[:, 0]) * tf.gradients(y, x)[0][:, 0]` have the same execution time if `tf.gradients(y, x)[0]` is precomputed, see the test script
```python
import numpy as np
import tensorflow as tf
import time


def test(simple: bool):
    tf.compat.v1.disable_eager_execution()

    data = np.random.rand(10 ** 8, 3)
    x = tf.compat.v1.placeholder(float, shape=(None, 3))
    y = tf.reduce_sum(x, axis=-1)
    dy_dx = tf.gradients(y, x)[0]

    result_simple = tf.gradients(tf.sin(x[:, 0]) * y, x)[0][:, 0]
    result_expand = (
        tf.gradients(tf.sin(x[:, 0]), x)[0][:, 0] * y
        + tf.sin(x[:, 0]) * tf.gradients(y, x)[0][:, 0]
    )

    sess = tf.compat.v1.Session()

    sess.run(dy_dx, feed_dict={x: data})

    print("=" * 10)
    if simple:
        result = result_simple
        print("simple")
    else:
        result = result_expand
        print("complex")

    t_start = time.time()
    sess.run(result, feed_dict={x: data})
    t_diff = time.time() - t_start
    print(f"time used: {t_diff:.3f}")
    sess.close()

    print("=" * 10)


if __name__ == "__main__":
    test(True)
    test(False)
```

I do observe expand all the term can get shorter execution time, but I am not sure the cause of it.

I am curious about that because sometimes the term in PDE is very complex and one do want to avoid expand every term to dy_i/dx_j and d^2y_i/dx_j,dx_k, however I am worrying that directly using for example
```
dde.gradients.jacobian(f(x, y) * y * dy/dx, x, i, j)
```
is slower than expand every term.

Please kindly provide some suggestions, thank you very much!

**Comment by lululxvi** [comment]
You are right. When computing `d (sin(x)*y) / dx`, TensorFlow is not smart enough to reuse the result `dy/dx`. Even in a simpler case, if you compute `dy/dx` twice, TensorFlow will not reuse `dy/dx` computed in the first time. That is one motivation of `dde.gradients.jacobian`, so that we will reuse what we have already computed.

Or even simpler, if you just compute `c1 = a + b; c2 = a + b`, `c2` will not reuse the result in `c1`. `c1 = a + b; c2 = a + b` is twice the cost of `c1 = a + b`.

Manual expansion might be worth, because you need to do it once. Or you can use Mathematics to do it.

**Comment by smao-astro** [comment]
Hi Lu,

I see, I eventually chose to expand every term, thank you very much!
